### ISSUE TITLE
Examples not used allreduce ？

### ISSUE THREAD

**Issue by hmilxin** [uber/horovod#141]
HI ：
I tested the sample “tensorflow_mnist.py” 。
run：
mpirun -np 2 -H localhost:2 -bind-to none -map-by slot -x NCCL_DEBUG=INFO -x LD_LIBRARY_PATH ...

Allreduce is not used here, then the parameters are how to update it ? 
In our project , one note's result is similar to two nodes。 

Thanks!

**Comment by alsrgv** [comment]
@hmilxin, we just updated [tensorflow_mnist.py](https://github.com/uber/horovod/blob/master/examples/tensorflow_mnist.py) example to demonstrate best practices:

1. Scale the learning rate with the number of workers.
2. Scale down the number of steps with the number of workers.
3. Reduce learning rate, as it was too high.

It does use allreduce since the optimizer is wrapped in `hvd.DistributedOptimizer`.

**Comment by hmilxin** [comment]
@alsrgv thank you！
  I tested it, it's well.  But I found a new problem with “tensorflow_word2vec.py”  when running on two machines with 4 GPUs(). I have TF 1.4, horovod 0.11.1 and openmpi 3.0.0.
 Here is a command I use to run:  （use tcp）
   mpirun -np 4 \
      -H server1:2,server2:2 \
      -bind-to none -map-by slot \
      -x NCCL_DEBUG=INFO -x LD_LIBRARY_PATH \       
      python tensorflow_word2vec.py
  I found that the speed was very  slow ,,,       It's a lot slower than one node with 2 gpu。
 What's wrong with me， please help me ， ths！

**Comment by alsrgv** [comment]
@hmilxin, can you quantify how much time it takes between prints in single machine vs two machine case?

E.g., between these two:
```
Average loss at step  20000 :  5.79205642962
Average loss at step  22000 :  5.60272070277
```

Also, can you share which compilation flags did you use when you compiled Horovod?

**Comment by hmilxin** [comment]
@alsrgv 
mpirun -np 2 -H localhost:2 -bind-to none -map-by slot -x NCCL_DEBUG=INFO -x LD_LIBRARY_PATH python tensorflow_word2vec.py
Log：--------------------------------------------------------------
Average loss at step  0 :  269.390014648 
Average loss at step  0 :  299.440856934 
Average loss at step  2000 :  87.3621843588  step_Time: 12.083
Average loss at step  2000 :  86.4530362191  step_Time: 12.082
Average loss at step  4000 :  30.1505744072  step_Time: 12.130
Average loss at step  4000 :  30.077325371  step_Time: 12.130
Average loss at step  6000 :  16.7555473597  step_Time: 12.128
Average loss at step  6000 :  16.882514281  step_Time: 12.128
Average loss at step  8000 :  11.6109631708  step_Time: 12.088
Average loss at step  8000 :  11.7268373234  step_Time: 12.088
---------------------------------------------------------------------
And: mpirun -np 4 -H 172.25.81.244:2,172.25.81.248:2 -bind-to none -map-by slot -x NCCL_DEBUG=INFO -x LD_LIBRARY_PATH --mca btl_tcp_if_exclude lo python tensorflow_word2vec.py
Log:------------------------------------------------------------------------------
Average loss at step  0 :  264.418640137
Average loss at step  0 :  284.843200684
Average loss at step  0 :  244.939285278
Average loss at step  0 :  254.734353462
Average loss at step  2000 :  69.8045976262  step_Time: 24.578
Average loss at step  2000 :  69.8262222235  step_Time: 24.579
Average loss at step  2000 :  69.5396038291  step_Time: 24.580
Average loss at step  2000 :  69.5345343453  step_Time: 24.580
Average loss at step  4000 :  19.6797439247  step_Time: 24.455
Average loss at step  4000 :  19.6382948261  step_Time: 24.455
Average loss at step  4000 :  19.4564523545  step_Time: 24.455
Average loss at step  4000 :  19.1734497596  step_Time: 24.455
-----------------------------------------------------------------------------------

Compiled Horovod  used：
$ HOROVOD_GPU_ALLREDUCE=NCCL HOROVOD_GPU_ALLGATHER=MPI HOROVOD_GPU_BROADCAST=MPI pip install --no-cache-dir horovod

Did I need to install  SoftRoCE ？

**Comment by hmilxin** [comment]
@alsrgv  another question：
We needed “clip_by_global_norm”  to calculate the gradient， In Horovod， where to add more good then？

eg：  
tvars = tf.trainable_variables()
grads, _ = tf.clip_by_global_norm(tf.gradients(cost, tvars), config.max_grad_norm)
self._train_op = optimizer.apply_gradients(zip(grads, tvars), global_step=global_step)

**Comment by alsrgv** [comment]
@hmilxin, thanks for the timing information.  Basically, time just grows linearly - it is expected for that example.  The word2vec model is very small and is mostly used to demonstrate/test working of sparse gradients used by embeddings.  Time per iteration is very low - 12 sec per 2000 iterations, 6ms per step.  This leaves very little time to communicate gradients between workers.

In order to make scaling more efficient, you can try to increase batch size by 100x and adjust LR (probably increase 100x, too).

As for your gradient clipping question, you can do that with Horovod by replacing `tf.gradients()` with `optimizer.compute_gradients(cost, tvars)`

**Comment by hmilxin** [comment]
@alsrgv ,  thank you !!!      I will try again,  
Is horovod a plan for async mode?

**Comment by alsrgv** [comment]
@hmilxin, not at the moment.  A number of researchers indicated that appearance of speed up produced by async mode is often offset by stale gradient problem, and as result model not converging to comparable accuracy.

**Comment by alsrgv** [comment]
Closing this issue due to lack of activity.

**Comment by cklsoft** [comment]
> our gradient clipping question, you can do that with Horovod by replacing `tf.gradients()` with `optimizer.compute_gradients(cost, tvars)`

I use `tf.clip_by_global_norm(grads, clip_norm=1.0)` to clip to gradients. When run horovod distributed, I found it not workers not communicate with each other.  I followed you suggestion to replace `grads = tf.gradients(loss, tvars)` to 

    `grads_and_vars = optimizer.compute_gradients(loss, tvars);
      grads = [grad for grad,var in grads_and_vars]`

. But it still not workers. Here it my code:

`
optimizer = hvd.DistributedOptimizer(optimizer)
tvars = tf.trainable_variables()
grads_and_vars = optimizer.compute_gradients(loss, tvars)
(grads, _) = tf.clip_by_global_norm(grads, clip_norm=1.0)
train_op = optimizer.apply_gradients(zip(grads, tvars), global_step=global_step)
`

**Comment by alsrgv** [comment]
@cklsoft, can you elaborate on why 5 line code does not work?  Is it throwing an error, or simply not training?

**Comment by cklsoft** [comment]
> can you elaborate on why 5 line code does not work? Is it throwing an error, or simply not training?

Thanks for your reply.
From the `global_step/sec` Horovod printed is same as In single worker, and the network traffic is same as single worker. 
![image](https://user-images.githubusercontent.com/3956357/50620923-8d539e80-0f3d-11e9-9884-104a31c9e5d4.png)

I ran one worker or multi-workers, `global_step/sec` aways 4.0+.

**Comment by alsrgv** [comment]
@cklsoft, could you paste code a bit more precisely?
```python
  optimizer = hvd.DistributedOptimizer(optimizer)
  tvars = tf.trainable_variables()
  grads_and_vars = optimizer.compute_gradients(loss, tvars)
  (grads, _) = tf.clip_by_global_norm(grads, clip_norm=1.0)
  train_op = optimizer.apply_gradients(zip(grads, tvars), global_step=global_step)
```

In this code, there's no declaration of `grads`.

**Comment by cklsoft** [comment]
> grads

@alsrgv Sorry for the missing.

          optimizer = hvd.DistributedOptimizer(optimizer)
          tvars = tf.trainable_variables()
          grads_and_vars = optimizer.compute_gradients(loss, tvars)
          grads = [grad for grad,var in grads_and_vars]
         (grads, _) = tf.clip_by_global_norm(grads, clip_norm=1.0)
          train_op = optimizer.apply_gradients(zip(grads, tvars), global_step=global_step)

**Comment by alsrgv** [comment]
@cklsoft, I'm not sure if this is the issue for your model, but passing explicit trainable variables is not the same as None that minimize() method is passing, which translates to:
```python
variables.trainable_variables() + ops.get_collection(ops.GraphKeys.TRAINABLE_RESOURCE_VARIABLES))
```

If you print out `tvars`, does it have all the variables you want to train?

If so, could you collect [Horovod timeline](https://github.com/uber/horovod/blob/master/docs/timeline.md) to determine if any averaging is happening?

**Comment by cklsoft** [comment]
@alsrgv 
Yes, You are right.  I modified as following:


      optimizer = hvd.DistributedOptimizer(optimizer)
      #tvars = tf.trainable_variables()
      grads_and_vars = optimizer.compute_gradients(loss) #no set tvars
      grads = [grad for grad,var in grads_and_vars]
      tvars = [var for grad,var in grads_and_vars]
      (grads, _) = tf.clip_by_global_norm(grads, clip_norm=1.0)
      train_op = optimizer.apply_gradients(zip(grads, tvars), global_step=global_step)

But now I found that, this code is twice faster than using `train_op = optimizer.minimize(loss=loss,global_step=tf.train.get_global_step())`. Bellow is the logs:

Using `minimize`:
![image](https://user-images.githubusercontent.com/3956357/50629202-8eed8880-0f76-11e9-87c0-c989b24a1362.png)

Using `apply_gradients`:
![image](https://user-images.githubusercontent.com/3956357/50629233-b04e7480-0f76-11e9-9286-05725d519052.png)

From the timeline, I found that the `minimize` method need 6ms for `ALLREDUCE`:
![image](https://user-images.githubusercontent.com/3956357/50629277-d8d66e80-0f76-11e9-9480-cb2dd2d2201f.png)

, But the `apply_gradients` method only need 3ms.
![image](https://user-images.githubusercontent.com/3956357/50629342-23f08180-0f77-11e9-8301-2b7cb976b7b5.png)

What is the different between two methods?

**Comment by alsrgv** [comment]
@cklsoft, could you collect Horovod timeline for both methods for hundred steps?  That way we'll be able to pinpoint the difference precisely.

**Comment by cklsoft** [comment]
> @cklsoft, could you collect Horovod timeline for both methods for hundred steps? That way we'll be able to pinpoint the difference precisely.

I have collected these timelines which are 80MB+.
[Archive.zip](https://github.com/uber/horovod/files/2725767/Archive.zip)

**Comment by alsrgv** [comment]
@cklsoft, the other difference is with `global_step`. Can you try passing `tf.train.get_global_step()` to `apply_gradients()`, like you do to `minimize()`?

**Comment by cklsoft** [comment]
> @cklsoft, the other difference is with `global_step`. Can you try passing `tf.train.get_global_step()` to `apply_gradients()`, like you do to `minimize()`?

The problem has been sloved. Thank you! 👍
